src/decoder/ifd.rs

```diff
@@ -10,27 +10,41 @@ use {TiffError, TiffFormatError, TiffResult, TiffUnsupportedError};
 use self::Value::{Ascii, List, Rational, Unsigned, Signed, SRational};
 
 macro_rules! tags {
-    {$(
-        $tag:ident
-        $val:expr;
-    )*} => {
-
-        /// TIFF tag
+    {
+        // Permit arbitrary meta items, which include documentation.
+        $( #[$enum_attr:meta] )*
+        $vis:vis enum $name:ident($ty:ty) $(unknown($unknown_doc:literal))* {
+            // Each of the `Name = Val,` permitting documentation.
+            $($(#[$ident_attr:meta])* $tag:ident = $val:expr,)*
+        }
+    } => {
+        $( #[$enum_attr] )*
         #[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
-        pub enum Tag {
-            $($tag,)*
-            Unknown(u16)
+        pub enum $name {
+            $($(#[$ident_attr])* $tag,)*
+            // FIXME: switch to non_exhaustive once stabilized and compiler requirement new enough
+            #[doc(hidden)]
+            __NonExhaustive,
+            $( 
+                #[doc = $unknown_doc]
+                Unknown($ty),
+            )*
         }
-        impl Tag {
-            pub fn from_u16(n: u16) -> Tag {
-                $(if n == $val { Tag::$tag } else)* {
-                    Tag::Unknown(n)
+
+        impl $name {
```

Comment:
I'm coming to this somewhat late so perhaps I'm missing some context, but is there a reason not to replace the methods below with impls of `TryFrom<$ty> for $tag` and `From<$tag> for $ty`?

Reply:
I thought I'd minimize the public interface that comes from *within* the macro. It might still be a good idea to expose those.

Reply:
We'll get `#[non_exhaustive]` (see https://github.com/rust-lang/rust/issues/44109) in a few releases and when we upgrade the compile requirements then we can definitely think of reworking this to purely perform discriminant casting as appropriate.